ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol/HotKey.hs

```diff
@@ -130,54 +187,82 @@ mkHotKey initKey startPeriod@(Absolute.KESPeriod start) maxKESEvolutions = do
             -- We always start from 0 as the key hasn't evolved yet.
           , kesEvolution   = 0
           }
-      , kesStateKey = initKey
+      , kesStateKey = KESKey initKey
       }
 
 -- | Evolve the 'HotKey' so that its evolution matches the given KES period.
 --
--- When the 'HotKey' has already evolved further than the given KES
--- period, we do nothing.
+-- When the given KES period is after the end period of the 'HotKey', we
+-- poison the key and return 'KESEvolutionFailed'.
+--
+-- When the given KES period is before the start period of the 'HotKey' or
+-- when the given period is before the key's period, we don't evolve the key
+-- and return 'KESUnchanged'.
+--
+-- When the given KES period is within the range of the 'HotKey' and the given
+-- period is after the key's period, we evolve the key and return
+-- 'KESEvolved'.
 --
--- When the given KES period is outside the bounds of the 'HotKey', we
--- also do nothing.
+-- When trying to evolve a poisoned key, we return 'KESEvolutionFailed'.
 --
--- In both cases, the 'checkKesPeriod' check performed on the 'KESInfo' in
--- 'checkIsLeader' will tell that the key is unusable, which means that we
--- will not forge a block and thus not try to sign anything using the key.
+-- NOTE: when the key is poisoned, but we don't try to evolve it, we return
+-- 'KESUnchanged'.
```

Comment:
Return `KESKeyPoisoned`, a new constructor of `KESEvolutionError`. I don't know when this would occur, but if it did, we'd still try to sign a block, causing an `error`.

Update the docstring: when updating a poisoned key, you *always* get `KESKeyPoisoned`.